storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java

```diff
@@ -253,31 +254,53 @@ public void onPartitionLeadershipChanges(Set<TopicIdPartition> leaderPartitions,
         Objects.requireNonNull(leaderPartitions, "leaderPartitions can not be null");
         Objects.requireNonNull(followerPartitions, "followerPartitions can not be null");
 
+        log.info("Received leadership notifications with leader partitions {} and follower partitions {}",
+                 leaderPartitions, followerPartitions);
+
+        HashSet<TopicIdPartition> allPartitions = new HashSet<>(leaderPartitions);
+        allPartitions.addAll(followerPartitions);
         lock.readLock().lock();
         try {
-            ensureInitializedAndNotClosed();
-
-            log.info("Received leadership notifications with leader partitions {} and follower partitions {}",
-                    leaderPartitions, followerPartitions);
-
-            HashSet<TopicIdPartition> allPartitions = new HashSet<>(leaderPartitions);
-            allPartitions.addAll(followerPartitions);
-            for (TopicIdPartition partition : allPartitions) {
-                remotePartitionMetadataStore.maybeLoadPartition(partition);
+            if (closing.get()) {
+                throw new IllegalStateException("This instance is in closing state");
             }
 
-            consumerManager.addAssignmentsForPartitions(allPartitions);
+            if (!initialized.get()) {
+                // If it is not yet initialized, then keep them as pending partitions and
+                this.pendingAssignPartitions = allPartitions;
+            } else {
+                this.pendingAssignPartitions = null;
```

Comment:
When this condition will be triggered? Assume that all the leadership notifications are received before initialization.